File: medium/number_of_islands.py
void dfs(vector<vector<char>>& grid, int i, int j, vector<vector<bool>>& check) {
        check[i][j] = 1;
        vector<int> x = {-1, 1, 0, 0};
        vector<int> y = {0, 0, 1, -1};
        for(int l = 0; l < 4; l++) {
            if(i+x[l] < 0 || i+x[l] >= grid.size())
                continue;
            if(j+y[l] < 0 || j+y[l] >= grid[0].size())
                continue;
            if(grid[i+x[l]][j+y[l]] == '1' && check[i+x[l]][j+y[l]] == 0)
                dfs(grid, i+x[l], j+y[l], check);
        }
    }
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if(grid.size() == 0) return 0;
        vector<vector<bool>> check(grid.size(), vector<bool>(grid[0].size(), 0));
        int count = 0;
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[0].size(); j++) {
                if(grid[i][j] == '1' && check[i][j] == 0) {
                    count++;
                    dfs(grid, i, j, check);
                }
            }
        }
        return count;
    }
};

File: medium/number_of_islands.cpp
class Solution {
public:
    int total_islands = 0;
    void dfs(vector<vector<char>>& grid, vector<vector<bool>>& visited, int i, int j) {
        visited[i][j] = 1;
        int directions[2] = {-1, 1};
        for(auto &x: directions) {
            if(i+x >= 0 && i+x < grid.size()) {
                if(grid[i+x][j] == '1' && !visited[i+x][j])
                    dfs(grid, visited, i+x, j);
            }
            if(j+x >= 0 && j+x < grid[0].size()) {
                if(grid[i][j+x] == '1' && !visited[i][j+x])
                    dfs(grid, visited, i, j+x);
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        if(grid.empty()) return total_islands;
        int m = grid.size(), n = grid[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i][j] == '1' && !visited[i][j]) {
                    total_islands++;
                    dfs(grid, visited, i, j);
                }
            }
        }
        return total_islands;
    }
    // Time Complexity:- O(n*m)
    // Space Complexity:- O(n*m)
};
